### PR TITLE
Lazy load data source data for list UI components

### LINKED ISSUE #37190: Duplicated calls on all Listing uiComponents
### Preconditions and environment

- Magento version 2.4.x
- Anything else that would help a developer reproduce the bug

### Steps to reproduce

open any admin listing page that is built using uiCompnennt,  e.g. customers listing page

### Expected result

one call to dataProvider to fetch the customers on the list

### Actual result

2 calls to dataProvider.

The first call occurs in the dom when initializing the uiComponent
The 2nd call occurs in ajax with default sorting etc.

### Additional information

I believe this creates a big performance issue as it pretty much duplicates most of the queries on page load for admin. The issue also gets worse if dataProvider involves slow query or API. 
IMO we should eliminate the call to dataprovider for data collection when initializing the component, and just let it being handled by the AJAX calls.

### Release note

_No response_

### Triage and priority

- [X] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by ryansunxl**
### Description (*)
If a page has listing UI component, the data source collection of the component is loaded and added to the page DOM in the generated component js tag. However this data collection is never rendered and always refreshed via ajax immediately after page load. It is waste of resource and slows down initial page load. 

### Fixed Issues (if relevant)

1. Fixes magento/magento2#37190

### Manual testing scenarios (*)

1. open any admin listing page that is built using uiCompnennt, e.g. customers listing page

### Questions or comments

I think this is a quite exciting performance improvement for many admin pages that has listing component. In my local testing, the TTFB imporved from about 570ms - 420ms

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by kuafucode** [comment]
@magento run all tests

File: app/code/Magento/Ui/Component/Listing.php [hunk 1]
```diff
@@ -35,6 +35,13 @@ public function getComponentName()
...
     {
-        return ['data' => $this->getContext()->getDataProvider()->getData()];
+        // Only load data in an ajax (json) request for better performance
+        if($this->getContext()->getAcceptType() == 'json') {
+            return ['data' => $this->getContext()->getDataProvider()->getData()];
+        }
+        else {
+            // build out default array keys to revent null erros, in case some module expects this
+            return ['data' => ['items' => [], 'totalRecords' => 0]];
+        }
```
Comment (ihor-sviziev) [inline app/code/Magento/Ui/Component/Listing.php]:
```suggestion
        // Only load data in an ajax (json) request for better performance
        if ($this->getContext()->getAcceptType() == 'json') {
            return ['data' => $this->getContext()->getDataProvider()->getData()];
        }

        // build out default array keys to revent null erros, in case some module expects this
        return ['data' => ['items' => [], 'totalRecords' => 0]];
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
I suggest removing else condition for better readability.
Please review and fix failing tests

**Comment by ihor-sviziev** [comment]
Hi @kuafucode,
Will you be able to fix failing tests?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please review and fix failing tests

**Comment by kuafucode** [comment]
the test is not running and not showing which test case failed, let me try again

**Comment by kuafucode** [comment]
@magento run all tests

**Comment by kuafucode** [comment]
Was able to locate the failed test case, the current PR breaks 'recently viewed' widget. In stead of making the change to `Magento\Ui\Component\Listing\Columns\Listing`, I should move this change into lower component `Magento\Framework\View\Element\UiComponent\DataProvider\DataProvider`.

I'm also thinking of introducing a lazy loading flag to data provide class via ui xml, instead of relying content format in request which is not semantic.

ps. I'm a bit suprised to find out the  'recently viewed' widget config is provided by `Magento\Catalog\Ui\DataProvider\Product\Listing\DataProvider` , IMO it can be fully provided in the ui xml config section

**Comment by engcom-Dash** [comment]
> Was able to locate the failed test case, the current PR breaks 'recently viewed' widget. In stead of making the change to `Magento\Ui\Component\Listing\Columns\Listing`, I should move this change into lower component `Magento\Framework\View\Element\UiComponent\DataProvider\DataProvider`.
> 
> I'm also thinking of introducing a lazy loading flag to data provide class via ui xml, instead of relying content format in request which is not semantic.
> 
> ps. I'm a bit suprised to find out the 'recently viewed' widget config is provided by `Magento\Catalog\Ui\DataProvider\Product\Listing\DataProvider` , IMO it can be fully provided in the ui xml config section

Hello @kuafucode,

I wanted to follow up on your comment regarding the changes needed for the 'recently viewed' widget and the lazy loading flag. Will you be implementing these updates? If not, please let us know how we should proceed.

Thanks!

**Comment by engcom-Dash** [comment]
Closing this PR since it has not been updated for a period of 14 days. Please feel free to reopen it if needed.

**Comment by kuafucode** [comment]
@engcom-Dash please reopen, I'm picking up this work again.

**Comment by engcom-Dash** [comment]
> @engcom-Dash please reopen, I'm picking up this work again.

@kuafucode- I am reopening this PR. Please review and address the  https://github.com/magento/magento2/pull/37576#issuecomment-1810885747 so we can proceed. Thank you!

**Comment by engcom-Dash** [comment]
@kuafucode: As discussed over Slack, I'm putting it on hold for now. Once you are done with the implementation, please let us know. We'll pick it up and proceed with the necessary steps. Thanks.

**Comment by engcom-Dash** [comment]
Hello @kuafucode,

This is a reminder to please address this https://github.com/magento/magento2/pull/37576#issuecomment-2418967790.

Thanks!

**Comment by engcom-Charlie** [comment]
Hi @kuafucode,

Did you get a chance to look into [this](https://github.com/magento/magento2/pull/37576#issuecomment-2418967790).

Thank you!

**Comment by engcom-Charlie** [comment]
Hi @kuafucode,

Did you get a chance to look into this [comment](https://github.com/magento/magento2/pull/37576#issuecomment-2418967790)?

Thank you!

**Comment by engcom-Charlie** [comment]
Hi @ryansunxl,

Did you get a chance to look into [this](https://github.com/magento/magento2/pull/37576#issuecomment-2418967790) change. Currently I am closing this PR, please feel free to reopen it once done with the required implementation.
